In [145]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

In [146]:
# write the linear regression hypothesis function in vector form
def hypothesis(feature, weight): 
    return np.dot(weight, feature)

predict = hypothesis(feature, weight)

In [147]:
# create a function to calculate the loss function in vector form
def loss(predict, target): 
    return 1 / (2 * len(feature)) * np.sum((predict - target) ** 2)

initial_loss = loss(predict, target)

In [148]:
# implement one step of gradient descent
alpha = 0.01
gradient = np.dot(feature, (predict - target)) / len(feature)
weight_new = weight - alpha * gradient

In [149]:
df = pd.read_csv('Housing.csv')
df.head()

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

In [150]:
X = df[['area', 'bedrooms', 'bathrooms']].values
y = df[['price']].values

In [151]:
# find the best $\vec{w}$ parameters for the dataset predicting house price depending on square footage, number of bathrooms, and number of bedrooms
regressor = LinearRegression()
regressor.fit(X, y)
w_pred = regressor.coef_

In [152]:
# find the same parameters using an analytical solution
w_analytic = np.dot(np.dot(np.linalg.inv(np.dot(X.T, X)), X.T), y)

In [153]:
# compare the results obtained
difference = w_pred.ravel() - w_analytic.ravel()
percentage_difference = (difference / w_analytic.ravel()) * 100
percentage_difference

array([ 1.69537653, 10.25689978,  1.14837633])

The percentage difference between the weights obtained analytically and using scikit-learn:

For the first weight: 1.70%
For the second weight: 10.26%
For the third weight: 1.15%

A small percentage difference may be acceptable and may arise from different optimization methods or numerical aspects of the calculation.